tiledb/sm/buffer/buffer.h

```diff
@@ -44,18 +43,91 @@ namespace sm {
 
 class ConstBuffer;
 
-/** Enables reading from and writing to a buffer. */
-class Buffer {
+// =======================================================
+//      BufferBase
+// =======================================================
+/**
+ * Base class for `Buffer` and `ConstBuffer`
+ *
+ * Responsible for maintaining an read offset in the range [0..size_]. Not
+ * responsible for memory management.
+ */
+class BufferBase {
  public:
-  /* ********************************* */
-  /*     CONSTRUCTORS & DESTRUCTORS    */
-  /* ********************************* */
+  /** Returns the buffer size. */
+  uint64_t size() const;
+
+  /** Returns the current read position, in bytes. */
+  uint64_t offset() const;
+
+  /** Resets the buffer offset to 0. */
+  void reset_offset();
+
+  /** Sets the buffer offset to the input offset. */
+  void set_offset(uint64_t offset);
+
+  /** Advances the offset by *nbytes*. */
+  void advance_offset(uint64_t nbytes);
+
+  /** Checks if reading has reached the end of the buffer. */
```

Comment:
While reading at end is primary use (and probably made sense with original 'ConstBuffer', I don't see 'end()' in other buffer classes), 
now technically just checking if 'current position' is at end of allocated buffer for any descendants of BufferBase?

Reply:
Changed the comment.